### ISSUE TITLE
Finish Python 2 removal

### ISSUE THREAD

**Issue by Carreau** [scikit-image/scikit-image#3117]
Followup on @jni pining me on #3000 (really good PR number for this BTW).

One of the critical things to do is to add the `python_requires='>=3.4'` to setup.py and add some extra failsafes in `setup.py` to ensure users get a good error message if they have an old pip version. Read http://python3statement.org/practicalities/ or more info.

Do you want to add scikit image to https://github.com/python3statement/python3statement.github.io ?

**Comment by jni** [comment]
@Carreau Yes, as noted in the PR description, I "nabbed" the # when I saw it coming. =D

Thanks for the link — I was unaware of the `python_requires` option. I knew pinging you was a good idea. =D

It feels a bit redundant to sign the statement now. I would have liked to sign it in the first wave but there was some resistance. However it is perfectly aligned with our roadmap (which is not yet written down but I intend to solve that this week), so yes, we can add scikit-image to it.

**Comment by stefanv** [comment]
LTS support for 2.7 on the 0.14.x branch will end 1 January 2020.  The first 3-only release will be 0.15.0, around November 2018–January 2019.

**Comment by stefanv** [comment]
@Carreau The instructions on python3statement are a bit confusing: if you specify `python_requires`, that would stop newer pips also on 2.7 from installing.  If you throw an error in `setup.py` if 2.7 is detected, what happens to pip?  So, an example of how to fail in setup.py would also be useful.  For now, I'll add the `python_requires` and throw an error on skimage import if on 2.

**Comment by Carreau** [comment]
Thanks for your question, 

It is a really good one, and the crux of why it took us about a year to get things right and patch upstream PyPI. We need to take a step back and realize that setup.py has 2 functions:

 - Build the package on the maintainer machine, and release it
 - Build the package on the user machine and install it (whl, or tgz). 

The `requires_python` set a metadata on the tgz/whl so that pip 9+ will not _attempt_ to download if the current Python is Python 2.7 (or 3.0, 3.1...). 

Though if you do _not_ have pip 9+ installed, then:
 1. pip will see the `py3.whl` extension
 2. decide there is no wheels for Py2, thus fallback on the `.tgz`.
 3. It will then unpack the `.tgz` and try to install the package; in which case the `python_requires` is ignored by setuptools. 

Now you have two choice. 
  - Either your `setup.py` is not Python 2 compatible, in which case the install with crash with an underfined error (SyntaxError, TypeError... whatever make setup.py not Python 2 compatible). But better what you _can_ do is fail on purpose with an explicit error message "No, Scikit image is not Python 2 compatible, please update pip, and/or install `<X.Y`" explicitly. 
 - You "succeed" at installing a Python 2 incompatible version, and will fail at runtime. This is suboptimal as then `pip install --upgrade` may break a perfectly working user install, and may only be detected days later. 


Note, `install skimage<X.Y` is also important in the message and not `skimage==X.Y` because you _may_ decide to make point release. We did make the first ipython 6.0 error message say `please install IPython==5.0` and now have a non-negligible amount of user that are not getting any of the `5.x` fixes as they pinned to `5.0` instead of `<6.0`. 

TLDR:
 - don't try to download whl/tgz on recent pip (`requires_python`)
 - if old pip:
   - fail at install time with **informative error message of what to do next** and **leave user Python install in a working state**. 


Does that clarify a bit why ?

**Comment by Carreau** [comment]
See https://github.com/python3statement/python3statement.github.io/pull/128 as well.

**Comment by stefanv** [comment]
Thank you; this explains the (rather horrifying) situation!
